### ISSUE TITLE
[tibber] monthly and annual information.

### ISSUE THREAD

**Issue by B4dR4cc00n** [openhab/openhab-addons#16797]
Please add the following channels to the binding for getting monthly and annual information.

- monthly_cost
- monthly_consumption
- annual_cost
- annual_consumption

**Comment by DrScr3w** [comment]
Hello, is it possible to include this in the next release please?

**Comment by lsiepel** [comment]
> Hello, is it possible to include this in the next release please?

If these are available by the API it should not be that hard to implement. Do you have the JSON or schema details? If these are not available, it might be better to calculate this using rules and persistence.

**Comment by DrScr3w** [comment]
@lsiepel 
I hope that I have determined it correctly ;)
https://developer.tibber.com/explorer

Under “Consumptions” you can see the “cost” and "consumption" and there you can switch from “hourly” to “monthly” or "annual".
![image](https://github.com/user-attachments/assets/ce5763e9-07e9-4c94-8c85-0a8c8447da38)
`{
  viewer {
    homes {
      consumption(resolution: MONTHLY, last: 100) {
        nodes {
          from
          to
          cost
          unitPrice
          unitPriceVAT
          consumption
          consumptionUnit
        }
      }
    }
  }
}`

![image](https://github.com/user-attachments/assets/8e22b564-bfd1-4a4f-8b4d-288660008f71)
![image](https://github.com/user-attachments/assets/97d5109e-6942-4f00-9037-e6caf05f0ca6)

**Comment by DrScr3w** [comment]
Wouldn't it “only” be necessary to add more IFs with monthly and annual?

https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java

Line 234
```
if (jsonResponse.contains("daily") && !jsonResponse.contains("\"daily\":{\"nodes\":[]")
        && !jsonResponse.contains("\"daily\":null")) {
    try {
        JsonObject myObject = (JsonObject) rootJsonObject.getAsJsonObject("data").getAsJsonObject("viewer")
                .getAsJsonObject("home").getAsJsonObject("daily").getAsJsonArray("nodes").get(0);

        String timestampDailyFrom = myObject.get("from").toString().substring(1, 20);
        updateState(DAILY_FROM, new DateTimeType(timestampDailyFrom));

        String timestampDailyTo = myObject.get("to").toString().substring(1, 20);
        updateState(DAILY_TO, new DateTimeType(timestampDailyTo));

        updateChannel(DAILY_COST, myObject.get("cost").toString());
        updateChannel(DAILY_CONSUMPTION, myObject.get("consumption").toString());
    } catch (JsonSyntaxException e) {
        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                "Error communicating with Tibber API: " + e.getMessage());
    }
}
```

**Comment by lsiepel** [comment]
You are right, it is not a complicated change. But there is more to it: the README.md and the thing-structure.xml need to be adapted and don't forget the thing upgrade instructions.

**Comment by weymann** [comment]
@DrScr3w 
> I hope that I have determined it correctly ;) https://developer.tibber.com/explorer
> 
> Under “Consumptions” you can see the “cost” and "consumption" and there you can switch from “hourly” to “monthly” or "annual".

First I like this feature because it would cover (nearly) the complete Tibber data
- future day ahead pricing
- live data measure
- history to check aftermatch

Data fetching is correct.  You can query history data with yearly, monthly, weekly and daily granularity.  **_But only if the timeperiod is over!_** 
If you query now yearly consumption & cost this will end 31.12.2024. IWith updating the item state the value appears at 2025. And the 2025 data will be avialble at Jan 2026. 
Looking into the graph or access the item value in rules  it's shifted by one year (month, week or day) which is simply wrong.

To avoid that I'm creating now a `TimeSeries `for the past to have the correct timestamp! The graph is correct **_but_** the item state stays `NULL `-  also understandable because we've no actual data! This would raise too many questions and I still need a solution for that. 

https://github.com/user-attachments/assets/88359fae-1e2e-4023-9e20-df900a354b65

<img width="1120" height="193" alt="Image" src="https://github.com/user-attachments/assets/6911f235-84c5-4cde-a102-7d6c456b212c" />

**Comment by DrScr3w** [comment]
`Data fetching is correct. You can query history data with yearly, monthly, weekly and daily granularity. But only if the timeperiod is over!`

Does this also apply to the monthly values?

Because of HA integration, I receive the amount paid so far for the current month, which is still ongoing.

**Comment by weymann** [comment]
> Does this also apply to the monthly values?

The [query you showed earlier in your comment](https://github.com/openhab/openhab-addons/issues/16797#issuecomment-2455553459)  requesting MONTHLY values will stop in July, right?  At least it does for me so no partial value for August is given.

I can only think about more complex queries to calculate the partial monthly consumption & cost out of the daily values.

**Comment by DrScr3w** [comment]
Yes, with the simple query, I can only see the completed months or years.
```
{
  viewer {
    homes {
      consumption(resolution: MONTHLY, last: 2) {
        nodes {
          from
          to
          cost
          unitPrice
          unitPriceVAT
          consumption
          consumptionUnit
        }
      }
    }
  }
}
```

```
"from": "2025-06-01T00:00:00.000+02:00",
"to": "2025-07-01T00:00:00.000+02:00",

"from": "2025-07-01T00:00:00.000+02:00",
"to": "2025-08-01T00:00:00.000+02:00",
```

Good question, how do they do that in HA Binding?
